test/integration/graphql_test.lua

```diff
@@ -564,25 +564,63 @@ g.test_custom_type_scalar_variables = function()
         package.loaded['test']['test_custom_type_scalar_list'] = function(_, args)
           return args.fields[1]
         end
+        package.loaded['test']['test_json_type'] = function(_, args)
+          if args.field == nil then
+            return nil
+          end
+          assert(type(args.field) == 'table', "Field is not a table! ")
+          assert(args.field.test ~= nil, "No field 'test' in object!")
+          return args.field
+        end
         package.loaded['test']['test_custom_type_scalar_inputObject'] = function(_, args)
           return args.object.nested_object.field
         end
 
+        local json = require('json')
         local graphql = require('cartridge.graphql')
         local types = require('cartridge.graphql.types')
 
+        local function isString(value)
+            return type(value) == 'string'
+        end
+
+        local function coerceString(value)
+            if value ~= nil then
+                value = tostring(value)
+                if not isString(value) then return end
+            end
+
+            return value
+        end
+
         local custom_string = types.scalar({
             name = 'CustomString',
             description = 'Custom string type',
-            serialize = tostring,
-            parseValue = tostring,
+            serialize = coerceString,
+            parseValue = coerceString,
+            parseLiteral = function(node)
+                return coerceString(node.value)
+            end,
+            isValueOfTheType = isString,
+        })
+
+        local function decodeJson(value)
+            if value ~= nil then
+                return json.decode(value)
+            end
+            return value
+        end
+
+        local json_type = types.scalar({
+            name = 'Json',
+            description = 'Custom type with JSON decoding',
+            serialize = json.encode,
+            parseValue = decodeJson,
             parseLiteral = function(node)
-              if node.kind == 'string' then
-                return node.value
-              end
+                return decodeJson(node.value)
             end,
-            isValueOfTheType = function(value)
-              return type(value) == 'string'
+            isValueOfTheType = function(_)
+                return true
```

Comment:
Seems it should be false for non-string values

Reply:
Yes, but in this case it does not matter